cgi-bin/DW/API/Key.pm

```diff
@@ -57,12 +57,17 @@ sub new_for_user {
 sub get_key {
     my ( $class, $hash ) = @_;
     return undef unless $hash;
-
-    my $dbh     = LJ::get_db_writer() or croak "Failed to get database";
-    my $keydata = $dbh->selectrow_hashref(
-        "SELECT keyid, userid, hash FROM api_key WHERE hash = ? AND state = 'A'",
-        undef, $hash );
-    carp $dbh->errstr if $dbh->err;
+    my $memkey = [ $hash, "api_key:" . $hash ];
+
+    my $keydata = LJ::MemCache::get($memkey);
+    unless ( defined $keydata ) {
+        my $dbh = LJ::get_db_writer() or croak "Failed to get database";
+        $keydata = $dbh->selectrow_hashref(
+            "SELECT keyid, userid, hash FROM api_key WHERE hash = ? AND state = 'A'",
+            undef, $hash );
+        LJ::MemCache::set( $memkey, $keydata, 60 * 60 * 24 );    # cache for one day
+        carp $dbh->errstr if $dbh->err;
```

Comment:
I would move this above the memcache set, so that we don't put invalid data into memcache.